File: interview-prep-kit/dictionaries_and_hashmaps/ransom_note/ransom_note.js
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', function () {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function createDictObj(arrOfWords) {
    let obj = {};
    for (let i = 0; i < arrOfWords.length; i++) {
        if (obj[arrOfWords[i]]) {
            obj[arrOfWords[i]] += 1;
        } else {
            obj[arrOfWords[i]] = 1;
        }
    }
    return obj;
}

// Complete the checkMagazine function below.
function checkMagazine(magazine, note) {
    let m = createDictObj(magazine);
    let n = createDictObj(note);

    for (let w in n) {
        if (m[w] < n[w] || !m[w]) {
            console.log("No");
            return;
        }
    }
    console.log("Yes");
}

function main() {
    const mn = readLine().split(' ');

    const m = parseInt(mn[0], 10);

    const n = parseInt(mn[1], 10);

    const magazine = readLine().split(' ');

    const note = readLine().split(' ');

    checkMagazine(magazine, note);
}

File: interview-prep-kit/dictionaries_and_hashmaps/ransom_note/RansomNote.java
import java.util.*;

public class RansomNote {

    // Complete the checkMagazine function below.
    static void checkMagazine(String[] magazine, String[] note) {
        HashMap<String, Integer> map = new HashMap<>();

        // populate hashmap with words in magazine
        for (int i = 0; i < magazine.length; i++) {
            if (map.containsKey(magazine[i])) {
                map.replace(magazine[i], map.get(magazine[i]) + 1);
            } else {
                map.put(magazine[i], 0);
            }
        }

        // check words from note
        for (int i = 0; i < note.length; i++) {
            if (map.containsKey(note[i])) {
                if (map.get(note[i]) < 0) {
                    System.out.println("No");
                    return;
                } else {
                    map.replace(note[i], map.get(note[i]) - 1);
                }
            } else {
                System.out.println("No");
                return;
            }
        }
        System.out.println("Yes");
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        String[] mn = scanner.nextLine().split(" ");

        int m = Integer.parseInt(mn[0]);

        int n = Integer.parseInt(mn[1]);

        String[] magazine = new String[m];

        String[] magazineItems = scanner.nextLine().split(" ");
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int i = 0; i < m; i++) {
            String magazineItem = magazineItems[i];
            magazine[i] = magazineItem;
        }

        String[] note = new String[n];

        String[] noteItems = scanner.nextLine().split(" ");
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int i = 0; i < n; i++) {
            String noteItem = noteItems[i];
            note[i] = noteItem;
        }

        checkMagazine(magazine, note);

        scanner.close();
    }
}
